component/file_cache/file_cache.go

```diff
@@ -704,107 +743,172 @@ func (fc *FileCache) OpenFile(options internal.OpenFileOptions) (*handlemap.Hand
+		err = fc.NextComponent().CopyToFile(
+			internal.CopyToFileOptions{
+				Name:   options.Name,
+				Offset: 0,
+				Count:  0,
+				File:   f,
+			})
+		if err != nil {
+			// File was created locally and now download has failed so we need to delete it back from local cache
+			log.Err("FileCache::download : error downloading file from storage %s [%s]", options.Name, err.Error())
+			_ = f.Close()
+			_ = os.Remove(localPath)
+			return nil, err
 		}
+	}
 
-		// Update the last download time of this file
-		flock.SetDownloadTime()
+	// Update the last download time of this file
+	flock.SetDownloadTime()
 
-		log.Debug("FileCache::OpenFile : Download of %s is complete", options.Name)
-		f.Close()
+	log.Debug("FileCache::download : Download of %s is complete", options.Name)
+	f.Close()
 
-		// After downloading the file, update the modified times and mode of the file.
-		fileMode := fc.defaultPermission
-		if attr != nil && !attr.IsModeDefault() {
-			fileMode = attr.Mode
-		}
+	// After downloading the file, update the modified times and mode of the file.
+	fileMode := fc.defaultPermission
+	if attr != nil && !attr.IsModeDefault() {
+		fileMode = attr.Mode
+	}
+
+	// If user has selected some non default mode in config then every local file shall be created with that mode only
+	err = os.Chmod(localPath, fileMode)
+	if err != nil {
+		log.Err("FileCache::download : Failed to change mode of file %s [%s]", options.Name, err.Error())
+	}
+	// TODO: When chown is supported should we update that?
 
-		// If user has selected some non default mode in config then every local file shall be created with that mode only
-		err = os.Chmod(localPath, fileMode)
+	if attr != nil {
+		// chtimes shall be the last api otherwise calling chmod/chown will update the last change time
+		err = os.Chtimes(localPath, attr.Atime, attr.Mtime)
 		if err != nil {
-			log.Err("FileCache::OpenFile : Failed to change mode of file %s [%s]", options.Name, err.Error())
+			log.Err("FileCache::download : Failed to change times of file %s [%s]", options.Name, err.Error())
 		}
-		// TODO: When chown is supported should we update that?
+	}
 
-		if attr != nil {
-			// chtimes shall be the last api otherwise calling chmod/chown will update the last change time
-			err = os.Chtimes(localPath, attr.Atime, attr.Mtime)
-			if err != nil {
-				log.Err("FileCache::OpenFile : Failed to change times of file %s [%s]", options.Name, err.Error())
-			}
-		}
+	fileCacheStatsCollector.UpdateStats(stats_manager.Increment, dlFiles, (int64)(1))
 
+	// Open the file and grab a shared lock to prevent deletion by the cache policy.
+	f, err = common.OpenFile(localPath, options.Flags, options.Mode)
+	if err != nil {
+		log.Err("FileCache::download : error opening cached file %s [%s]", options.Name, err.Error())
+		return nil, err
+	}
+
+	// Increment the handle count in this lock item as there is one handle open for this now
+	flock.Inc()
+
+	handle := handlemap.NewHandle(options.Name)
```

Comment:
If you pass the handle into `DownloadFile`, you could update the struct fields by simply skipping this line.
Then the lines below would update the existing handle instead of creating a new one.